LeanOA/ForMathlib/Analysis/SpecialFunctions/ContinuousFunctionalCalculus/Abs.lean

```diff
@@ -0,0 +1,236 @@
+  simpa [NNReal.abs_eq] using abs_smul_real r a
+
+end Real
+
+section Complex
+
+variable [NonUnitalRing A] [TopologicalSpace A] [Module ℂ A]
+
+variable [StarRing A] [PartialOrder A] [StarOrderedRing A]
+variable [IsScalarTower ℂ A A] [SMulCommClass ℂ A A]
+variable [NonUnitalContinuousFunctionalCalculus ℂ (IsStarNormal : A → Prop)]
+
+open ComplexOrder in
+lemma cfcₙ_norm_sq_nonneg {f : ℂ → ℂ} {a : A} : 0 ≤ cfcₙ (fun z ↦ star (f z) * (f z)) a :=
+  cfcₙ_nonneg fun _ _ ↦ star_mul_self_nonneg _
+
+open ComplexOrder in
+lemma cfcₙ_norm_nonneg {a : A} : 0 ≤ cfcₙ (fun z : ℂ ↦ (‖z‖ : ℂ)) a :=
+  cfcₙ_nonneg fun _ _ ↦ by simp only [Complex.zero_le_real, norm_nonneg]
+
+variable [NonnegSpectrumClass ℝ A] [IsTopologicalRing A] [T2Space A]
+
+lemma abs_sq_eq_cfcₙ_norm_sq_complex (a : A) (ha : IsStarNormal a) :
+    abs a ^ (2 : NNReal) = cfcₙ (fun z : ℂ ↦ (‖z‖ ^ 2 : ℂ)) a := by
+  conv_lhs => rw [abs_nnrpow_two, ← cfcₙ_id' ℂ a, ← cfcₙ_star, ← cfcₙ_mul ..]
+  exact cfcₙ_congr fun x hx ↦ Complex.conj_mul' x
+
+lemma abs_eq_cfcₙ_norm_complex (a : A) (ha : IsStarNormal a) :
+    abs a = cfcₙ (fun z : ℂ ↦ (‖z‖ : ℂ)) a := by
+  conv_lhs => rw [abs, ← abs_nnrpow_two, sqrt_eq_real_sqrt .., cfcₙ_real_eq_complex,
+    abs_sq_eq_cfcₙ_norm_sq_complex a ha, ← cfcₙ_comp' ..]
+  exact cfcₙ_congr fun x hx ↦ by simp [sq]
+
+lemma cfcₙ_abs_complex (f : ℂ → ℂ) (a : A) (ha : IsStarNormal a := by cfc_tac)
+    (hf : ContinuousOn f ((fun z ↦ (‖z‖ : ℂ)) '' quasispectrum ℂ a) := by cfc_cont_tac) :
+    cfcₙ f (abs a) = cfcₙ (fun x ↦ f ‖x‖) a := by
+  rw [abs_eq_cfcₙ_norm_complex a ha]
+  obtain (hf0 | hf0) := em (f 0 = 0)
+  · rw [← cfcₙ_comp' ..]
+  · rw [cfcₙ_apply_of_not_map_zero _ hf0, cfcₙ_apply_of_not_map_zero _ (fun h ↦ (hf0 <| by simpa using h).elim)]
+
+variable [StarModule ℂ A] in
+lemma abs_smul_complex (r : ℂ) (a : A) : abs (r • a) = ‖r‖ • abs a := by
+  have : 0 ≤ ‖r‖ • abs a := smul_nonneg (by positivity) abs_nonneg
+  rw [abs, CFC.sqrt_eq_iff _ _ (star_mul_self_nonneg _) this]
+  simp only [mul_smul_comm, smul_mul_assoc, abs_mul_self, star_smul]
+  match_scalars
+  rw [mul_one, mul_one, ← sq, mul_comm, RCLike.star_def, Complex.conj_mul', ← Complex.coe_algebraMap]
+
+end Complex
+
+end NonUnital
+
+section Unital
+
+section Real
+
+variable [Ring A] [StarRing A] [PartialOrder A] [StarOrderedRing A] [TopologicalSpace A] [Algebra ℝ A] [IsTopologicalRing A] [T2Space A]
+variable [ContinuousFunctionalCalculus ℝ (IsSelfAdjoint : A → Prop)]
+variable [NonnegSpectrumClass ℝ A]
+
+@[simp]
+lemma abs_one : abs (1 : A) = 1 := by
+  rw [abs, star_one , mul_one, sqrt_one]
+
+variable [StarModule ℝ A]
+
+lemma abs_algebraMap_real (c : ℝ) : abs (algebraMap ℝ A c) = algebraMap ℝ A |c| := by
+  simp only [Algebra.algebraMap_eq_smul_one, abs_smul_real, abs_one]
+
+lemma abs_algebraMap_nnreal (x : ℝ≥0) : abs (algebraMap ℝ≥0 A x) = algebraMap ℝ≥0 A x := by
+  simpa only [NNReal.abs_eq] using abs_algebraMap_real (NNReal.toReal _)
```

Comment:
This could be generalized like `abs_smul_nonneg`, but I think it *shouldn't* be, since the simp normal form of `algebraMap ℕ A n` is `Nat.cast n : A`.
```suggestion
@[simp]
lemma abs_algebraMap_nnreal (x : ℝ≥0) : abs (algebraMap ℝ≥0 A x) = algebraMap ℝ≥0 A x := by
  simp [Algebra.algebraMap_eq_smul_one]
```